doc/gui/extension/example_library/front-end/src/BasicTable.tsx

```diff
@@ -0,0 +1,149 @@
+import React, { useEffect, useMemo, useState, useCallback, ReactNode } from "react";
+import {
+    createRequestDataUpdateAction,
+    RowType,
+    RowValue,
+    useDispatch,
+    useDispatchRequestUpdateOnFirstRender,
+    useDynamicProperty,
+    useModule,
+} from "taipy-gui";
+
+interface BasicTableProps {
+    id?: string;
+    updateVarName?: string;
+    updateVars?: string;
+    data: Record<string, Record<string, unknown>>;
+    rowsPerPage?: number;
+}
+
+const BasicTable = (props: BasicTableProps) => {
+    const [value, setValue] = useState<Record<string, unknown>>({});
+    const [currentPage, setCurrentPage] = useState(1);
+    const dispatch = useDispatch();
+    const module = useModule();
+    const refresh = props.data?.__taipy_refresh !== undefined;
+    useDispatchRequestUpdateOnFirstRender(dispatch, props.id, module, props.updateVars, props.updateVarName);
+
+    // Memoize column order and columns
+    const [colsOrder, columns] = useMemo(() => {
+        const colsOrder = Object.keys(value || {}).sort();
+        return [colsOrder, value || {}];
+    }, [value]);
+
+    // Memoize rows
+    const rows = useMemo(() => {
+        const rows: RowType[] = [];
+        if (value) {
+            colsOrder.forEach(
+                (col) =>
+                    value[col] &&
+                    (value[col] as RowValue[]).forEach(
+                        (val, idx) => (rows[idx] = rows[idx] || {}) && (rows[idx][col] = val),
+                    ),
+            );
+        }
+        return rows;
+    }, [value, colsOrder]);
+
+    // Memoize paginated rows
+    const paginatedRows = useMemo(() => {
+        if (props.rowsPerPage === undefined) {
+            return rows; // Show all rows if itemsPerPage is undefined
+        }
+        const itemsPerPageValue = props.rowsPerPage ?? 10;
+        const startIndex = (currentPage - 1) * itemsPerPageValue;
+        return rows.slice(startIndex, startIndex + itemsPerPageValue);
+    }, [rows, currentPage, props.rowsPerPage]);
+
+    // Update data callback
+    const updateData = useCallback(() => {
```

Comment:
why declare a useCallback when the code can be put directly in the useEffect ?

Reply:
resolved